some SPI refactorization

vchong opened the pull request:
A bit many commits as they're separated into digestible pieces for easier reviewing.

Replaces #1215 
Fixes #1208 

@gitfineon @daniel-thompson fyi

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -157,18 +159,26 @@
 		return;
 	}
 
+	if (pd->data_size_bits == 16) {
```
Wouldn't `if (pd->data_size_bits != 8)` be better? (and same below for the 16-bit version).

vchong replied:
Yes you're right. Will change.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -195,13 +213,18 @@ static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat, uint16_t *rdat,
 		return;
 	}
 
+	if (pd->data_size_bits == 8) {
```
`!= 16` instead?

vchong replied:
Ok. Will change.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -184,6 +194,14 @@ static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat, uint8_t *rdat,
+			else
+				cnt++;
+
+			if (cnt > SSP_LOOP_TIMEOUT) {
```
How can this be reliable on all platforms? (there is no timer or time function involved). Same for the 16-bit version below.

vchong replied:
@jforissier You're right! I tested on HiKey with what only seemed like a sufficiently big enough loop count. Time functions are not available and even if they are, it might not easy to find a time value that works for all platforms.

I've pushed a change that checks for register busy status instead, which is what I should have done in the first place, so thanks for catching this!

vchong commented:
@daniel-thompson 

> Why does num_rxpkts need an initializer.

From https://github.com/OP-TEE/optee_os/pull/905#r69943362, num_rxpkts is supposed to hold the number of max packets to receive. If not initialized, there is the number of loops in txr8 will be undefined/random. In any case, num_rxpkts doesn't matter anymore as it is removed in a later [patch](https://github.com/OP-TEE/optee_os/pull/1251/commits/4ac8ec3071c726a121e54d2d64df14050c80db62).

vchong commented:
Comments addressed and checkpatch warnings/errors fixed.

jforissier commented:
@vchong I think this looks OK, could you please squash the fixup commits where they belong for a final review?

vchong commented:
@jforissier Thanks for not forgetting this! :D
I've squashed the fixup commits.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -123,8 +123,10 @@
+#define SSPICR_RTIC		SHIFT_U32(1, 1)
+#define SSPICR_RTIC_CLEAR	SHIFT_U32(1, 1)
+#define SSPICR_RORIC		SHIFT_U32(1, 0)
+#define SSPICR_RORIC_CLEAR	SHIFT_U32(1, 0)
```
minor: seems useless duplications of `SSPICR_RTIC` and `SSPICR_RORIC`, these are already 'clear signal' bit fields.

vchong replied:
Ok. Will remove `SSPICR_RTIC_CLEAR` and `SSPICR_RORIC_CLEAR`.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -180,11 +188,17 @@ static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat, uint8_t *rdat,
+
+		if (j < num_pkts) {
+			EMSG("# packets requested: %zu\n", num_pkts);
+			EMSG("# packets received: %zu\n", j);
```
This looks more informative message than an error message since nothing fails (no error status propagated). If `pl022_txrx8()` can fail, shouldn't there rather be a return value for this? 
Same for `pl022_txrx16()` below.

minor: rather trace this over a single string line.

vchong replied:
Ok. Will use `IMSG` over a single string line.

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -180,11 +188,17 @@ static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat, uint8_t *rdat,
+		if (j < num_pkts) {
+			EMSG("# packets requested: %zu\n", num_pkts);
+			EMSG("# packets received: %zu\n", j);
+		}
```
Better on one line:
```C
                if (j < num_pkts)
                        EMSG("Packets requested %zu, received %zu", num_pkts, j);
```

vchong replied:
Ok. Will also change `EMSG` to `IMSG` per Etienne's comment.

vchong replied:
Checkpatch says this is too long. :( Change it back?

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -202,6 +216,11 @@ static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat, uint16_t *rdat,
 
+	if (pd->data_size_bits != 16) {
+		EMSG("data_size_bits should be 16, not %u\n",
+			pd->data_size_bits);
```
Should fit on the previous line

vchong replied:
ok

vchong replied:
Checkpatch says this is too long. :( Change it back?

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -169,6 +171,12 @@ static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat, uint8_t *rdat,
 
+	if (pd->data_size_bits != 8) {
+		EMSG("data_size_bits should be 8, not %u\n",
+			pd->data_size_bits);
```
Should fit on the previous line

vchong replied:
ok

vchong replied:
Checkpatch says this is too long. :( Change it back?

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -214,11 +233,17 @@ static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat, uint16_t *rdat,
+		if (j < num_pkts) {
+			EMSG("# packets requested: %zu\n", num_pkts);
+			EMSG("# packets received: %zu\n", j);
+		}
```
Better on one line

vchong replied:
Ok. Will also change `EMSG` to `IMSG` per Etienne's comment.

vchong replied:
Checkpatch says this is too long. :( Change it back?

etienne-lms commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -48,45 +72,57 @@
 		DMSG("SPI test loop: %zu\n", j);
+
+		/* If cs_cb = NULL, CS can be enabled here */
+
```
minor: are these comments useful as start and end handlers from spi driver already controls CS ?

vchong replied:
These comments follow the one above that says "NULL can be registered, in which case you can pulse CS as desired below". They're for the use case where start and end handlers don't control the CS, i.e. when `b_cs_callback` is true.

Perhaps I was trying to show too much in one function. The other option is to have 3 spi test functions where:
i. `b_cs_callback` is false, i.e. start and end handlers don't control the CS
ii. `b_cs_callback' is true and `cs_cb` is not NULL
iii. `b_cs_callback' is true and `cs_cb` is NULL
Maybe this way it can illustrate better how the code differs between use cases?

jforissier commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,8 +28,32 @@
+#define PL022_STAT		0x00C
+#define PL022_STAT_BSY	SHIFT_U32(1, 4)
+
+static __maybe_unused void spi_cs_callback(enum gpio_level value)
```
`__maybe_unused` is probably not needed

jforissier commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -48,45 +72,57 @@
 void spi_test(void)
 {
-	struct pl061_data platform_pl061_data;
+	struct pl061_data platform_pl061_data __maybe_unused;
```
`__maybe_unused` is probably not needed

jforissier commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -48,45 +72,57 @@
+	struct pl061_data platform_pl061_data __maybe_unused;
 	struct pl022_data platform_pl022_data;
-	vaddr_t gpio6_base = nsec_periph_base(GPIO6_BASE);
+	vaddr_t __maybe_unused gpio6_base = nsec_periph_base(GPIO6_BASE);
```
`__maybe_unused` is probably not needed

jforissier commented on core/drivers/pl022_spi.c:
```diff
@@ -273,10 +265,11 @@ static void pl022_sanity_check(struct pl022_data *pd)
-	assert(pd->base);
-	assert(pd->cs_gpio_base);
+	/* pd->cs_data.cs_cb CAN be NULL so no assert */
+	if (!pd->b_cs_callback) {
```
Why do you need `b_cs_callback`? It ssems redundant with `cs_data.cs_cb`.

etienne-lms replied:
Because of the union in the structure, can't distinguish between 
Maybe the union is useless. Or maybe a "default" callback for acting on a gpio could be used so that is follows the same scheme! already call the CS callback handler.

vchong replied:
@jforissier Because registering NULL as `cs_data.cs_cb` is a valid use case, so there's no way to know if the caller wants to use `cs_data.cs_cb` or `gpio_data` just by checking `if (cs_data.cs_cb)`, so need a flag to indicate that.

@etienne-lms As above comment in `spi_test.c`, there are 3 use cases:
i. `b_cs_callback` is false, then `pl022_control_cs` will need `cs_data.gpio_data` info to control CS.
ii. `b_cs_callback` is true and `cs_cb` is not NULL, then `pl022_control_cs` will just call `cs_data.cs_cb` to control CS.
iii. `b_cs_callback` is true and `cs_cb` is NULL, then `pl022_control_cs` will do nothing, and let caller control CS manually however they want.

So for `cs_data`, it can only be either `gpio_data` or `cs_cb`, which is the reason an union is used. As in reply to Jerome above, `b_cs_callback` is used to distinguish between the two, because `cs_cb` alone cannot be used to know what caller wants due to use case of allowing registration of NULL as cb.

etienne-lms replied:
What about proposing a default callback for gpio based CS ?
This would make the code always go though a callback.
If my proposal simplifies the registration and executive sequence, it's fine. If it adds complexity, just forget it.

vchong replied:
The gpio based CS is sort of the default callback, just that in order to use it, caller has to fill in `gpio_data` or else the driver doesn't know which gpio pin to use.

jforissier commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -48,45 +72,57 @@
 	uint8_t tx[3] = {0x01, 0x80, 0x00};
 	uint8_t rx[3] = {0};
 	size_t i, j, len = 3;
+	bool local_cs = false;
```
Ah, now I see the reason for `__maybe_unused`. Make this `local_cs` a parameter to the test function instead. Then `spi_test()` could call `_spi_test(false);`, and possibly also `_spi_test(true);` to test both modes?

vchong replied:
As mentioned to @etienne-lms earlier, maybe it's better and clearer to just have 3 different spi_test functions to highlight the different use cases instead of trying to put everything into one.

jforissier commented on core/include/drivers/pl022_spi.h:
```diff
@@ -34,17 +34,26 @@
+	unsigned int		mode;
+	unsigned int		data_size_bits;
+	bool			b_cs_callback;
+	bool			b_loopback;
```
Why rename `loopback` to `b_loopback`?

vchong replied:
Because I added `b_cs_callback` and tried to keep both consistent. Since we don't use camel case like `bLoopback`, would you prefer just `cs_callback` and `loopback` instead?

jforissier replied:
I'd rather have `b_cs_callback` and `loopback` (because `cs_callback` would be ambiguous, could be the pointer itself, no such issue with `loopback`).

vchong replied:
ok for above and checkpatch warning fix below

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -158,155 +159,86 @@
+	if (pd->data_size_bits != 16) {
+		EMSG("data_size_bits should be 16, not %u\n",
+			pd->data_size_bits);
+		return;
```
is this an real error case ?
if so, there should be some error status propagation.

same for pl022_`tx`rx8().

vchong replied:
Ok. Will use `IMSG` over a single string line.

vchong replied:
Np. Will add return value.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -158,155 +159,86 @@
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+		if (j < num_pkts) {
+			EMSG("# packets requested: %zu\n", num_pkts);
+			EMSG("# packets received: %zu\n", j);
```
Same comment as above: is this an real error case ?
if so, there should be some error status propagation.

same for pl022_`tx`rx8().

vchong replied:
Ok. Will use `IMSG` over a single string line.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -356,6 +289,21 @@ static inline uint32_t pl022_calc_freq(struct pl022_data *pd,
+		pd->cs_data.cs_cb(value);
+	} else if (!pd->b_cs_callback) {
+		while (read8(pd->base + SSPSR) & SSPSR_BSY)
+			DMSG("pl022 still transceiving - do NOT set CS yet!");
```
this debug message lloks strange... is it really useful ?

vchong replied:
This is a case where SW is faster than the IP HW. If we set the pinmux function for the CS pin as SPI, then the IP HW will just control the CS pin and there's no problem there, but if we set the function as GPIO, then this driver is responsible for controlling/pulsing/changing the CS signal. In this case, if we just change the signal without checking to see if the IP HW is busy (SSPSR_BSY), i.e. still transmitting/receiving, then we will cancel the transfer in the middle of transmission. The debug message lets us know that program execution is at the point where it's ready to change the CS signal, but the HW IP is still busy so we need to wait.

etienne-lms replied:
I fear a trace overflow when the situation occurs.
If you feel this message is useful, maybe you could trace the debug message only once and add a "no more busy" message once busy signal cleared.

vchong replied:
Ah, ok, I see your point.
Will add an `is_busy` flag to do what you suggested.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -77,7 +77,7 @@ void spi_test(void)
 	platform_pl022_data.speed_hz = SPI_10_KHZ;
 	platform_pl022_data.cs_gpio_pin = GPIO6_2;
-	platform_pl022_data.mode = SPI_MODE0;
+	platform_pl022_data.mode = 0;
```
Mode0 is 0, mode1 is 1 and so on so I thought it was a bit redundant and could be simplified to look less cluttered. Would you rather bring back the defines?

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -158,153 +157,69 @@
-			/* tx 1 packet */
-			write8(wdat[i++], pd->base + SSPDR);
+	if (!num_pkts) {
+		EMSG("*num_pkts = 0!\n");
```
Ok. Will remove.

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -158,153 +157,69 @@
-		if (read8(pd->base + SSPSR) & SSPSR_TNF)
-			/* tx 1 packet */
-			write16(wdat[i++], pd->base + SSPDR);
+	if (wdat) {
```
ok

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -158,153 +157,69 @@
-			/* tx 1 packet */
-			write16(wdat[i++], pd->base + SSPDR);
+	if (wdat) {
+		while (i < num_pkts) {
```
ok

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -158,153 +157,69 @@
-		if (read8(pd->base + SSPSR) & SSPSR_TNF)
-			/* tx 1 packet */
-			write8(wdat[i++], pd->base + SSPDR);
+	if (rdat) {
```
Not sure why some of the comments are regarded as outdated? Did I force a push while you guys were in the middle of adding comments? Apologies if that was what happened!

Anyway, this brace is needed because an additional `if` statement is added in a later commit.

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -158,153 +157,69 @@
-static void pl022_tx16(struct spi_chip *chip, uint16_t *wdat,
-	size_t num_txpkts)
+static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat, uint16_t *rdat,
+	size_t num_pkts)
```
ok

jforissier commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -33,10 +33,18 @@
+
+	for (i = 0; i < (mult*10000000); i++)
+		;
+}
```
You can't rely on this to provide a "sleep at least this long" semantic. You should probably use `void tee_time_wait(uint32_t milliseconds_delay)`, which sleeps in normal world, or if it doesn't work for some reason then implement `void tee_time_mdelay(uint32_t ms)` which would actively wait but use `tee_time_get_sys_time()` to determine the exit condition rather than a hardcoded loop count.

vchong replied:
I will `tee_time_mdelay` in a separate commit by itself below. Please check if it's ok. Thanks!

etienne-lms commented on core/include/spi.h:
```diff
@@ -30,30 +30,18 @@
 
 #include <types_ext.h>
 
```
Regarding magic numbers and macros, yes i would personally prefer macros as they explicitly state the supported values in the source code. I agree execution if not affected since allowed values are verified.

vchong replied:
Ok. Macros added back.

vchong commented:
> Comment on commit message of commit "drivers: spi: simplify code": the comment refer of a pull request which may not be reachable someday (PR squashed, rebased, ...). I think you should rather explicitly detail the conversion conclusion

I don't intend to squash all the main (non-fixup) commits, but will also add the explicit detail you suggested to the relevant commit message before the final push.

Thanks!

jforissier commented on core/include/kernel/tee_time.h:
```diff
@@ -38,5 +38,6 @@ TEE_Result tee_time_get_ta_time(const TEE_UUID *uuid, TEE_Time *time);
 TEE_Result tee_time_get_ree_time(TEE_Time *time);
 TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time);
 void tee_time_wait(uint32_t milliseconds_delay);
+void tee_time_mdelay(uint32_t ms);
```
Please add /* Busy wait */ before the declaration to make it clear how it's different from `tee_time_wait()`.

vchong replied:
ok

jforissier commented on core/arch/arm/kernel/tee_time.c:
```diff
@@ -57,6 +57,17 @@ void tee_time_wait(uint32_t milliseconds_delay)
 	thread_rpc_cmd(OPTEE_MSG_RPC_CMD_SUSPEND, 1, &params);
 }
 
+void tee_time_mdelay(uint32_t ms)
```
Move it to `core/tee/tee_time_generic.c` since it's not arch-specific.

vchong replied:
ok

vchong commented:
All round 3 comments addressed. Please continue review. Thanks!

jforissier commented on core/arch/arm/kernel/tee_time.c:
```diff
@@ -57,6 +57,17 @@ void tee_time_wait(uint32_t milliseconds_delay)
+
+	do {
+		tee_time_get_sys_time(&curr);
+	} while (curr.millis - prev.millis < ms);
```
You need to consider seconds also! The macros `TEE_TIME_ADD()` and `TEE_TIME_LT()` may be helpful here.

vchong replied:
Yes, was researching the overflow case when you replied. The macros were indeed helpful. Thanks!

vchong commented:
Comments for `tee_time_mdelay` addressed.

jforissier commented:
For the `tee_time_mdelay()` patch, please use this subject: `core: Add tee_time_mdelay()`
Other than that: `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

vchong commented:
Ok, thanks. Will change subject before final push.

@etienne-lms Are you ok with changes?

etienne-lms commented on core/include/kernel/tee_time.h:
```diff
@@ -38,5 +38,7 @@ TEE_Result tee_time_get_ta_time(const TEE_UUID *uuid, TEE_Time *time);
 TEE_Result tee_time_get_ree_time(TEE_Time *time);
 TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time);
 void tee_time_wait(uint32_t milliseconds_delay);
```
maybe we should add a comment stating `tee_time_wait()` releases CPU trhough OP-TEE RPC means. It would help understanding the difference with `tee_time_mdelay()`.
Or maybe `tee_time_mdelay()` should be renamed `tee_time_busy_wait()`.

jforissier replied:
+1 for the comment on `tee_time_wait()`, it's important to know it switches to Normal World.

I originally proposed `tee_time_mdelay()` because the Linux kernel has `mdelay()`, `udelay()` and `ndelay()` for busy wait loops. I also like `tee_time_busy_wait()` altough it won't allow us to later implement microsecond or nanosecond delays and keep a consistent naming.

etienne-lms replied:
Let's cut the pear in two (french expression :) : what about `tee_busy_mdelay()` ?

jforissier replied:
@etienne-lms I don't like it :stuck_out_tongue_closed_eyes:

Let's use `tee_time_busy_wait()` because it's purpose is clear and it's consistent with our current naming scheme and usage of millisecond units.

vchong replied:
Will change to `tee_time_busy_wait()`.

vchong replied:
Also will addd comment for `tee_time_wait`.

etienne-lms commented on core/tee/tee_time_generic.c:
```diff
@@ -139,3 +139,21 @@ TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time)
+
+void tee_time_mdelay(uint32_t ms)
+{
+	TEE_Time prev, curr;
```
minor: prefer 1 variable define per line (even if of same type).

vchong replied:
Will change as @jforissier's proposed.

etienne-lms commented on core/tee/tee_time_generic.c:
```diff
@@ -139,3 +139,21 @@ TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time)
+
+		if (curr.seconds == prev.seconds)
+			ms_elapsed = curr.millis - prev.millis;
+		else
```
if `(curr.seconds > prev.seconds)` break now.
if `(curr.seconds < prev.seconds)` simply loop again.

jforissier replied:
What? I don't get it.
And BTW the proposed implementation is wrong too, as it won't be able to handle `ms > 2000` in any case (not sure we'd like to do some busy wait for more that 2 seconds but unless we explicitly check for a maximum value it should work). @vchong I withdraw my R-b on this patch -- sorry for that.
Here is my proposal:
```C
void tee_time_mdelay(uint32_t ms)
{
        TEE_Time curr, delta, end;

        tee_time_get_sys_time(&curr);
        delta.seconds = ms / 1000;
        delta.millis = ms % 1000;
        TEE_TIME_ADD(curr, delta, end);

        while (TEE_TIME_LT(curr, end))
                tee_time_get_sys_time(&curr);
}
```

vchong replied:
Sorry, np, and thanks!
Will update.

etienne-lms commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,236 @@
+	platform_pl022_data.base = spi_base;
+	platform_pl022_data.clk_hz = SPI_CLK_HZ;
+	platform_pl022_data.speed_hz = SPI_10_KHZ;
+	platform_pl022_data.mode = 0;
```
SPI_MODE_0

vchong replied:
will change

etienne-lms commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,236 @@
-	platform_pl022_data.cs_gpio_pin = GPIO6_2;
-	platform_pl022_data.mode = SPI_MODE0;
+	platform_pl022_data.speed_hz = SPI_10_KHZ;
+	platform_pl022_data.mode = 0;
```
SPI_MODE0

vchong replied:
will change

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -416,44 +356,50 @@ static void pl022_flush_fifo(struct pl022_data *pd)
-		DMSG("SPI_MODE0");
-		mode = PL022_SPI_MODE0;
+		DMSG("SPI mode 0");
+		mode = SSPCR0_SPO0 | SSPCR0_SPH0; /* 0x00 */
```
`mode` is set but not used.

vchong replied:
Comments removed.

`mode` is used in `io_mask16` below.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -463,11 +409,11 @@ void pl022_configure(struct pl022_data *pd)
 	case 8:
 		DMSG("Data size: 8");
-		data_size = PL022_DATA_SIZE8;
+		data_size = SSPCR0_DSS_8BIT;
```
`data_size` is set but not used.

vchong replied:
`data_size` is used in `io_mask16` below.

etienne-lms replied:
true. my apologies.

vchong commented:
> Well, a whole bunch of comments. Sorry, please do not feel try to delay the integration. And it's quite my fault since i did not wrote some of these comments earlier.

Np. Better to get it right the first time.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,236 @@
+	pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
+	platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
+						GPIO_INTERRUPT_DISABLE);
+	platform_pl061_data.chip.ops->set_direction(GPIO6_2, GPIO_DIR_OUT);
```
Is it ok to do below?

```
	static bool is_init = false;

	if (!is_init) {
		pl061_init(&platform_pl061_data);
		pl061_register(gpio6_base, 6);
		pl061_set_mode_control(GPIO6_2, PL061_MC_SW);
		platform_pl061_data.chip.ops->set_interrupt(GPIO6_2,
							GPIO_INTERRUPT_DISABLE);
		platform_pl061_data.chip.ops->set_direction(GPIO6_2, GPIO_DIR_OUT);
		is_init = true;
	}
```

Otherwise, if there's a separate init routine, we'll have to pass a pointer to `pl061_data` to the callback function, which will make it dependent on pl061 driver, which kinds of defeat the purpose of the callback, which is to make it flexible for the caller to implement a CS callback of their design.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,236 @@
+	platform_pl061_data.chip.ops->set_direction(GPIO6_2, GPIO_DIR_OUT);
+
+	while (read8(spi_base + PL022_STAT) & PL022_STAT_BSY)
+		DMSG("pl022 still transceiving - do NOT set CS yet!");
```
Will change as suggested below.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,236 @@
+	platform_pl022_data.base = spi_base;
+	platform_pl022_data.clk_hz = SPI_CLK_HZ;
+	platform_pl022_data.speed_hz = SPI_10_KHZ;
+	platform_pl022_data.mode = 0;
```
Yes, forgot to change this when adding the macros back. Will change. Thanks!

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,236 @@
+	platform_pl022_data.chip.ops->start(&platform_pl022_data.chip);
+
+	/* pulse CS only once for the whole transmission */
+	/* this is the scheme used by the pl022 driver */
```
ok

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -356,6 +271,31 @@ static inline uint32_t pl022_calc_freq(struct pl022_data *pd,
+{
+	bool is_busy = false;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
```
Yes, it's valid in SPI for users to not give GPIO data for CS and control CS GPIO pin manually. This allows them to use any GPIO pin they wish for CS.

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -356,6 +271,31 @@ static inline uint32_t pl022_calc_freq(struct pl022_data *pd,
+
+	if (pd->b_cs_callback && pd->cs_data.cs_cb) {
+		pd->cs_data.cs_cb(value);
+	} else if (!pd->b_cs_callback) {
```
Agree. Will change. Thanks!

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -356,6 +271,31 @@ static inline uint32_t pl022_calc_freq(struct pl022_data *pd,
+
+		if (is_busy) {
+			DMSG("pl022 done - set CS!\n");
+			is_busy = false;
```
Will change as suggested above.

vchong commented on core/drivers/pl022_spi.c:
```diff
@@ -497,25 +443,44 @@ void pl022_configure(struct pl022_data *pd)
-	pd->gpio->ops->set_direction(pd->cs_gpio_pin, GPIO_DIR_OUT);
+	DMSG("clear interrupts");
+	io_mask8(pd->base + SSPICR, (SSPICR_RORIC | SSPICR_RTIC),
+		(SSPICR_RORIC | SSPICR_RTIC));
```
Will remove.

vchong commented on core/include/drivers/pl022_spi.h:
```diff
@@ -29,27 +29,34 @@
 
+struct pl022_cs_gpio_data {
+	struct gpio_chip	*chip;
+	unsigned int		pin_num; /* chip select */
```
Will remove.

vchong commented:
All comments from yesterday addressed. Thanks for constructive reviews!

jforissier commented:
Please fix the checkpatch warnings/errors (the long URL is OK though), then:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,246 @@
+	for (j = 0; j < 10; j++) {
+		DMSG("SPI test loop: %zu\n", j);
+		platform_pl022_data.chip.ops->txrx8(&platform_pl022_data.chip,
+						tx, rx, len);
```
Should test the return value. Here and at several other places.
Same regarding `txrx16` handler.

vchong replied:
test code added

etienne-lms commented on core/include/drivers/pl022_spi.h:
```diff
@@ -29,27 +29,34 @@
+	vaddr_t			base;
+	unsigned int		clk_hz;
+	unsigned int		speed_hz;
+	unsigned int		mode;
```
i did not caught this one: `mode` should be of type `enum spi_mode`

vchong replied:
ok thanks

etienne-lms commented on core/tee/tee_time_generic.c:
```diff
@@ -139,3 +139,16 @@ TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time)
+
+void tee_time_busy_wait(uint32_t milliseconds_delay)
+{
+	TEE_Time curr, delta, end;
```
minor: split into 3 lines.

vchong replied:
ok

etienne-lms commented on core/tee/tee_time_generic.c:
```diff
@@ -139,3 +139,16 @@ TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time)
+{
+	TEE_Time curr, delta, end;
+
+	tee_time_get_sys_time(&curr);
```
check return value (type `TEE_Result`), here and below ?
should it panic on failure ? return an error ? ...

jforissier replied:
We can expect it will "never" fail, so IMO it's not worth returning an error status (and have have all call sites handle the status!). I'm OK with `if (res != TEE_SUCCESS) panic();` however.

etienne-lms replied:
@jforissier I agree with you but since `tee_time_get_sys_time()` returns a status, I think it should be either handled or explicitly discarded (hence the question marks in my comment).

vchong replied:
added `panic()`

vchong commented:
checkpatch errors also fixed
all recent comments addressed

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,279 @@
+
+static void spi_cs_callback(enum gpio_level value)
 {
+	static bool is_init = false;
```
Checkpatch says: ERROR: do not initialise statics to false.
Can this be ignored? Makes the code clearer with the initialization.

jforissier replied:
No, we should follow the checkpatch rules as much as possible.

etienne-lms replied:
@vchong, if you don't feel the boolean null value is not explicit enough as a 'false' identifier, maybe you can switch to a `static int inited;`: i think everyone expect such default initializes to 0.

jenswi-linaro replied:
Everyone should expect `static bool` to be initialized as `false` also.

vchong replied:
changed to `inited`

etienne-lms commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,279 @@
-						tx, rx, len, &num_rxpkts);
-		for (i = 0; i < num_rxpkts; i++)
+		res = platform_pl022_data.chip.ops->txrx8
+			(&platform_pl022_data.chip, tx, rx, len);
```
I would prefer the parenthesis to be right next to the handler (@jerome: your opinion)
```
		res = platform_pl022_data.chip.ops->txrx8(
					&platform_pl022_data.chip, tx, rx, len);
```
maybe you could find a way to prevent such line splits.

jforissier replied:
Yes it's better to keep the parenthesis on the same line. IMO you should `s/platform_pl022_data/pd/g` (or if you find `pd` a bit too cryptic use `pdata`).

vchong replied:
Changed to `pd`. No more splits. Thanks! Should have done that since the beginning.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -158,155 +157,78 @@
+		while (i < num_pkts)
+			if (read8(pd->base + SSPSR) & SSPSR_TNF)
+				/* tx 1 packet */
+				write8(wdat[i++], pd->base + SSPDR);
```
minor: could you add brackets here, around the comment + call to `write8()` ?
Same below for the 8bit receive.
Same below for the 16bit transmit and receive.

vchong replied:
done

etienne-lms commented on core/tee/tee_time_generic.c:
```diff
@@ -139,3 +140,19 @@ TEE_Result tee_time_set_ta_time(const TEE_UUID *uuid, const TEE_Time *time)
+	TEE_TIME_ADD(curr, delta, end);
+
+	while (TEE_TIME_LT(curr, end))
+		tee_time_get_sys_time(&curr);
```
same here: if routine (unlikely) fails: should panic.

vchong replied:
Right, missed that. Thanks!

vchong commented:
All comment addressed.

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -333,10 +259,11 @@ static void pl022_sanity_check(struct pl022_data *pd)
-	assert(pd->gpio->ops);
-	assert(pd->base);
-	assert(pd->cs_gpio_base);
+	/* pd->cs_data.cs_cb CAN be NULL so no assert */
```
Do you think `pd->cs_data.cs_cb` can really be `NULL` if `pd->b_cs_callback == true` ?
Is it a way to specify there is not Chip Select required ?

If so, I think you should explicitly state this in the structure definition, and I think it's a bit... ugly too. I would prefer to replace `b_cs_callback` with a enumerate field, like `cs = CS_GPIO` or `cs = CS_CALLBACK` or `cs = CS_NONE`.

vchong replied:
Ok. Will try an enum to see if it's better.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,271 @@
-	struct pl061_data platform_pl061_data;
-	struct pl022_data platform_pl022_data;
+	static bool inited;
+	struct pl061_data pd;
```
Sigh.. right, sorry, will change.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,271 @@
+	pd.data_size_bits = 8;
+	pd.loopback = true;
+
+	pl022_init(&pd);
```
Sorry I don't really get it? These functions are only called once.

etienne-lms replied:
ok if you are sure this will be called only once, as it is currently the case.
maybe those need a `xxx_deinit()` function. it is likely that some late code will init again the pl022 instance. I leave it up to you to decide.

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,271 @@
+	for (j = 0; j < 10; j++) {
+		DMSG("SPI test loop: %zu\n", j);
+		res = pd.chip.ops->txrx8(&pd.chip, tx, rx, len);
+
```
removed

vchong commented on core/arch/arm/plat-hikey/spi_test.c:
```diff
@@ -28,73 +28,271 @@
+		res = pd.chip.ops->txrx8(&pd.chip, tx, rx, len);
+
+		if (res) {
+			DMSG("SPI transceive error %d\n", res);
```
Changed

vchong commented:
Comments addressed

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -158,155 +157,82 @@
-	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
-
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+	if (wdat)
```
A comment that is out of scope of the PR: the SPI IP is likely to rely on some internal FIFO buffer of limited size. If one requests a exchange of more byte than the IP FIFO size, the IP will fail to buffer the full received SPI frame while data are sent, leading to some FIFO overflow.

Another one: if `rdat` is `NULL`, FIFO will not be wiped. Possible FIFO overflow issue. How will the IP behave ?
And another one,related: if `wdat` is NULL, IP won't transmit data: will the SPI communication operate in such condition? I think your test should treat those cases: only one of the `rdat` and `wdat` is `NULL`).

Am I right ? Should we open a Issue for that ? (this issue is out of scope of the PR, as former implementation used the same 'transmit all then receive all' scheme)

etienne-lms commented on core/drivers/pl022_spi.c:
```diff
@@ -356,6 +291,30 @@ static inline uint32_t pl022_calc_freq(struct pl022_data *pd,
+	case PL022_CS_CTRL_CB:
+		pd->cs_data.cs_cb(value);
+		break;
+	case PL022_CS_CTRL_MANUAL:
```
minor: usless `case PL022_CS_CTRL_MANUAL`. Leave it to `default`.

etienne-lms commented on core/include/drivers/pl022_spi.h:
```diff
@@ -29,27 +29,40 @@
 #define PL022_REG_SIZE	0x1000
 
+enum pl022_cs_control {
+	PL022_CS_CTRL_BUILTIN,
```
minor: `PL022_CS_CTRL_GPIO` is more explicit than `PL022_CS_CTRL_BUILTIN` as it is GPIO based.

vchong commented:
@etienne-lms I added commit to prevent possible FIFO overflow when one requests a exchange of more byte than the IP FIFO size.

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

vchong commented:
Thanks @etienne-lms!

Hi @jforissier Is your rb still ok?

jforissier commented:
> Hi @jforissier Is your rb still ok?

It is!

vchong commented:
Rebased, squashed and tags added. Some parts of commit messages have also been adjusted per comments/reviews/changes above to explain better what's actually happening in the patches.

Please merge if ok. Thanks! Phew!

jforissier commented:
Rebased on top of master and merged. Thanks @vchong!
